include/etcpal/cpp/error.h

```diff
@@ -187,8 +185,729 @@ inline bool operator!=(const Result& a, const Result& b)
   return !(a == b);
 }
 
+/// @}
+/// @}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+// etcpal::BadExpectedAccess
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+/// \brief Exception representing bad access to an Expected instance.
+/// \ingroup etcpal_cpp_error
+/// \details
+/// Thrown when attempting to access Expected<T>::value() when Expected<T>::has_value() is false.
+/// Holds an etcpal::Result; this allows nicer access to error information than just having the
+/// #etcpal_error_t code.
+class BadExpectedAccess : public std::exception
+{
+public:
+  explicit BadExpectedAccess(Result res) noexcept;
+  virtual const char* what() const noexcept override;
+  Result result() const noexcept;
+
+private:
+  Result res_;
+};
+
+/// \brief Construct from a Result.
+BadExpectedAccess::BadExpectedAccess(Result res) noexcept : res_(res)
+{
+}
+
+/// \brief Description of this exception type.
+const char* BadExpectedAccess::what() const noexcept
+{
+  return "Bad access to etcpal::Expected::value()";
```

Comment:
Would it be helpful to include the result in this to make it easier to see what happened if the exception just bubbles up and isn't caught explicitly?

Reply:
It would... building dynamic strings in exceptions is a bit tricky though because of all the noexcept functions. See [std::runtime_error](https://en.cppreference.com/w/cpp/error/runtime_error) which apparently has to implement a reference counted allocated string...

Reply:
Actually come to think of it, we could take advantage of that by having this class derive from runtime_error instead of exception...